interactions/api/http.py

```diff
@@ -0,0 +1,2285 @@
+        self.token = token
+        self.loop = get_event_loop()
+        self.session = session
+        self.ratelimits = {}
+        self.headers = {
+            "X-Ratelimit-Precision": "millisecond",
+            "Authorization": f"Bot {self.token}",
+            "User-Agent": f"DiscordBot (https://github.com/goverfl0w/discord-interactions {__version__} "
+            f"Python/{version_info[0]}.{version_info[1]} "
+            f"aiohttp/{http_version}",
+        }
+        self.lock = Event(loop=self.loop)
+        self.limit = 50  # representing all requests under 1 second.
+
+        self.lock.set()
+
+    def check_session(self) -> None:
+        """Ensures that we have a valid connection session."""
+        if self.session.closed:
+            self.session = ClientSession()
+
+    async def request(self, route: Route, **kwargs) -> Optional[Any]:
+        r"""
+        Sends a request to the Discord API.
+
+        :param route: The HTTP route to request.
+        :type route: Route
+        :param \**kwargs: Optional keyword-only arguments to pass as information in the request.
+        :type \**kwargs: dict
+        :return: The contents of the request if any.
+        :rtype: Optional[Any]
+        """
+        self.check_session()
+
+        bucket: Optional[str] = route.bucket
+
+        for _ in range(3):
+            ratelimit: Lock = self.ratelimits.get(bucket)
+
+            if not self.lock.is_set():
+                log.warning("Global lock is still locked, waiting for it to clear...")
+                await self.lock.wait()
+
+            if ratelimit is None:
+                self.ratelimits[bucket] = Lock()
+                continue
+
+            await ratelimit.acquire()
+
+            with Padlock(ratelimit) as lock:  # noqa: F841
+                kwargs["headers"] = {**self.headers, **kwargs.get("headers", {})}
+
+                try:
+                    reason = kwargs.get("reason")
+                except:  # noqa
+                    pass
+                else:
+                    if reason:
+                        kwargs["headers"]["X-Audit-Log-Reason"] = quote(reason, safe="/ ")
+
+                kwargs["headers"]["Content-Type"] = "application/json"
+                async with self.session.request(
+                    route.method, route.__api__ + route.path, **kwargs
+                ) as response:
+                    if not (response.headers["X-Ratelimit-Remaining"] - self.limit):
+                        log.info(
+                            "The application has reached the last remaining request. Cooling down to avoid rate limiting."
+                        )
+                        self.lock.set()
+                        await sleep(1)
+                        self.lock.clear()
+                    else:
+                        self.limit = response.headers["X-Ratelimit-Remaining"]
+                        data = await response.json(content_type=None)
+                        log.debug(data)
```

Comment:
Remove this, and replace it with logic that:

if `X-Ratelimit-Remaining` less than one
- If `X-Ratelimit-Global` or `X-Ratelimit-Scope` = 'global' (doesn't matter which)
-  - lock global lock and wait `X-RateLimit-Reset-After` milliseconds and unlock
- else `X-Ratelimit-Scope` = 'user'
- - lock per request ratelimit (this will be the `ratelimit` var and wait `X-RateLimit-Reset-After` milliseconds and unlock

Sorry about the *horrible* pseudocode, if you want i can put this into python but that'll take me a bit
here's the d.py implementation (yours should be a little different i think, just the way everything else has been setup) https://github.com/Rapptz/discord.py/blob/45d498c1b76deaf3b394d17ccf56112fa691d160/discord/http.py#L281

Reply:
```py
remaining = response.headers["X-Ratelimit-Remaining"]

if remaining == '0' and response.status != 429:

    time_left = ## Somehow parse `X-Ratelimit-Reset-After`
    # Lock the `ratelimit` (never used asyncio.Lock before so have fun :) )
    await asyncio.sleep(time_left)
    # Unlock `ratelimit`

```
I *think* discord won't return headers to avoid the global ratelimit, this is something to double check